### PR TITLE
Allow the lambda service to be used modularly

### PR DESCRIPTION
**Pull request by SeriousBug**
A useful feature for the CDK would be to allow the lambda service to be used modularly. For example, one might want to have multiple services that share the same API gateway. This would make it possible to break up a large monolithic service into smaller services, while keeping all the services within the same stack.

This PR achieves this by allowing the API gateway, API gateway stage, and API authorizers to be passed in as objects to the service. We still create these constructs if they are not provided so simpler setups don't have to bear the complexity of separately creating these constructs.

As an added improvement, this PR also encapsulates these now-modular constructs into their own objects. Apart from improving code readability, this allows someone creating a large multi-service stack to more easily create these constructs.

Finally, a small change in the PR allows `bundlingOptions` passed into the handlers to override the built-in source mapping settings. Some users of the cdk may want to disable source mapping as it can excessively increase bundle sizes.

---

All the changes in this PR are backwards compatible. The only major change is that the separate `jwtAuthorizer` and `lambdaAuthorizer` options are now deprecated in favor of a unified `authorizer` option. This is an improvement to the API as specifying `jwtAuthorizer` and `lambdaAuthorizer` together would result in one of them being ignored. Combining them into a single option more clearly communicates that only one of these can be used at a time. The old options are deprecated but still supported for backwards compatibility.

### CODE REVIEW

**Review by AnYoung627 — APPROVED** [review]
This looks just like what I had in mind! LMK how testing goes!

File: constructs/lambda-service.ts [hunk 1]
```diff
@@ -17,19 +17,47 @@ import { ServiceQueueFunction } from './service-queue-function';
...
+	 * gateway stage will be created.
+	 */
+	stage?: apigwv2.CfnStage;
+	/** The Authorizer to use for the API handlers.
+	 *
+	 * This can either be an actual authorizer, in which case we'll use it. Or it
+	 * can be a configuration for either a Jwt or Lambda authorizer, in which case
+	 * we'll create a new authorizer with that configuration.
+	 */
+	authorizer?: JwtAuthorizerConfig | LambdaAuthorizerConfig | CfnAuthorizer;
```
Comment (s-fletcher) [inline constructs/lambda-service.ts:60]: 🍟 Just for those who are unaware of this PR, we should keep `lambdaAuthorizer` and `jwtAuthorizer` but set their typing to `never` and have a `@deprecated` comment directing them to `authorizer`.

File: constructs/service-api-function.ts [hunk 2]
```diff
@@ -62,9 +62,9 @@ export class ServiceApiFunction extends Construct {
 			),
 			role: role,
 			bundling: {
-				...bundlingOptions,
 				sourceMap: true,
 				sourceMapMode: lambdaNodeJs.SourceMapMode.INLINE,
+				...bundlingOptions,
```
Comment (s-fletcher) [inline constructs/service-api-function.ts:67]: nice

File: constructs/lambda-service.ts [hunk 3]
```diff
@@ -55,6 +83,10 @@ export class LambdaService extends Construct implements iam.IGrantable {
 	readonly api: apigwv2.CfnApi;
 	readonly stage: apigwv2.CfnStage;
 	readonly grantPrincipal: iam.IPrincipal;
+	readonly authorizer?: apigwv2.CfnAuthorizer;
+
+	/** Maps queue names to the queue handlers of this service, if any. */
+	public queues: Map<string, ServiceQueueFunction> = new Map();
```
Comment (s-fletcher) [inline constructs/lambda-service.ts:89]: Out of curiosity, what's the application here?

**Review by s-fletcher — APPROVED** [review]
Nothing blocking. Only some 🍟

**Review by bill-kerr — APPROVED** [review]
LGTM. Don't have too much context here, but looks great

File: constructs/lambda-service.ts (hunk 1 above)
Comment (AnYoung627) [reply constructs/lambda-service.ts:60]: Great feedback, thanks @s-fletcher

File: constructs/lambda-service.ts (hunk 3 above)
Comment (SeriousBug) [reply constructs/lambda-service.ts:89]: If you want to access the queues of one service from another service, you have to explicitly grant the other service permissions. To do that, you need access to the queues of the service.

We can argue whether it's good for one service to directly access the queue of another service, but I think it's useful to make that possible if anyone needs it.

File: constructs/lambda-service.ts [hunk 4]
```diff
@@ -101,77 +134,43 @@ export class LambdaService extends Construct implements iam.IGrantable {
...
-				identitySource: [...lambdaAuthorizer.identitySource],
-				authorizerPayloadFormatVersion: '2.0',
-				enableSimpleResponses: lambdaAuthorizer.enableSimpleResponses,
 			});
-		} else if (jwtAuthorizer) {
-			authorizer = new apigwv2.CfnAuthorizer(this, 'JwtAuthorizer', {
-				apiId: this.api.ref,
-				authorizerType: 'JWT',
+		} else if (authorizer !== undefined) {
+			this.authorizer = new JwtAuthorizer(this, 'Authorizer', {
```
[the remark this answers is not in the document]
Comment (SeriousBug) [reply constructs/lambda-service.ts:160]: Not sure what you mean by this. The authorizer resource is only created if the config is provided. In the case where you are using multiple services, you provide the config for your first/root service then reuse the actual authorizer for other services.

Are you saying that you would need to recreate the authorizer? Or that the CDK would recreate it automatically even if you tried to reuse it?

**Pull request merged**